### ISSUE TITLE
Update “Most popular products by pageviews” key metric widget tile to use custom dimensions

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#7611]
## Feature Description

The “Most popular products by pageviews” key metric widget tile currently uses pattern matching to detect product content-type, which may prove to be unreliable in some cases. With the introduction of custom dimensions in Site Kit as a part of #7601, this tile should be updated to create and use a new `googlesitekit_post_type` custom dimension instead of pattern matching.

For product pages specifically, Site Kit should use `product` as the post type to track and report product content type, but this should also be filterable with a hook (via `apply_filters` for example) so that users or other plugins have the ability to define their own product post type.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- The “Most popular products by pageviews” key metric widget tile should be updated to use the `googlesitekit_post_type` custom dimension to determine product pages (instead of the existing product base paths usage).
  - The `googlesitekit_post_type` custom dimension should equal to a filterable (see [`apply_filters`](https://developer.wordpress.org/reference/functions/apply_filters/)) post type name that should default to `product`.
  - The product base paths implementation should be removed from the codebase if it is not being used elsewhere.

#### Backend changes

* In `includes/Core/Assets/Assets.php`
  * Add `get_product_post_type ` method: 
     * Apply a filter with the hook name `googlesitekit_product_post_type`, with the initial value being the string product (see [here](https://github.com/google/site-kit-wp/blob/00aa54ff148d37648ff6c7cf3189a8dbafefbf6f/includes/Modules/Analytics_4.php#L1406) for an example), and assign the result to a new variable, say `$product_post_type`
      * Include a check if `$product_post_type` is instance of post type and if the post type is public, like [here](https://github.com/google/site-kit-wp/blob/c9bd883549c11655029fae20f2ac5f3b4beb159f/includes/Core/Assets/Assets.php#L1088) 
        * If this condition is met, return `$product_post_type`, otherwise return `null`
  * Replace [productBasePaths](https://github.com/google/site-kit-wp/blob/00aa54ff148d37648ff6c7cf3189a8dbafefbf6f/includes/Core/Assets/Assets.php#L727) with new key say `productPostType` and for value invoke `get_product_post_type` method.
* In `includes/Modules/Analytics_4/Report/Request.php`
  * Add `googlesitekit_post_type` to the list of [valid dimensions](https://github.com/google/site-kit-wp/blob/d2c526086329e40e04eb875d1a1663e244e1c23f/includes/Modules/Analytics_4/Report/Request.php#L273-L289)

#### Report and datastore changes

* In `assets/js/googlesitekit/datastore/site/info.js`
  * include new `productPostType` variable in the list of [destructured object](https://github.com/google/site-kit-wp/blob/fb15edd0e2edb5bebbf6df94dd255ffc77c82173/assets/js/googlesitekit/datastore/site/info.js#L251-L275), and pass it to the `receiveSiteInfo`  in `getSiteInfo` resolver
  * Include `productPostType` in the [reducer](https://github.com/google/site-kit-wp/blob/fb15edd0e2edb5bebbf6df94dd255ffc77c82173/assets/js/googlesitekit/datastore/site/info.js#L138-L166), and pass it to the [state](https://github.com/google/site-kit-wp/blob/fb15edd0e2edb5bebbf6df94dd255ffc77c82173/assets/js/googlesitekit/datastore/site/info.js#L170-L198)
  * Add new selectors, like [this one](https://github.com/google/site-kit-wp/blob/00aa54ff148d37648ff6c7cf3189a8dbafefbf6f/assets/js/googlesitekit/datastore/site/info.js#L582) for example, say `getProductPostType` that will pull `productPostType` info property
  * Remove `productBasePaths` from the file
* Edit `assets/js/components/KeyMetrics/key-metrics-widgets.js`, under `KM_ANALYTICS_POPULAR_PRODUCTS` definition:
  * Add `requiredCustomDimensions` key and include `googlesitekit_post_type` as array value. You can see example [here](https://github.com/google/site-kit-wp/blob/a92636957ce393be24a10a7b03da7973e83e4df6/assets/js/components/KeyMetrics/key-metrics-widgets.js#L271)
  * Modify current check for `displayInList`, replace `getProductBasePaths` selector with previously added `getProductPostType` info store selector, which will be falsy if no product post type is present on the website.
* in `assets/js/modules/analytics-4/components/widgets/PopularProductsWidget.js`
  * Remove usage of `productBasePaths`
  * Use `getProductPostType` selector and assign the value to new variable
  * Update `reportOptions`
    * Add a `dimensionFilter` for the `customEvent:googlesitekit_post_type` dimension
    * Use `string` for `filterType`
    * Use `EXACT` for `matchType`
    * And `productPostType` as a value

### Test Coverage

* Remove usage of `productBasePaths` from `tests/phpunit/integration/Core/Assets/AssetsTest.php` and `tests/js/utils.js`, and `assets/js/googlesitekit/datastore/site/info.test.js`, also add `getProductPostType ` in info store tests
* Update `assets/js/modules/analytics-4/components/widgets/PopularProductsWidget.stories.js`
* Update any failing tests

## QA Brief

* Setup Site kit with Analytics module
  * Choose `News Key Metrics` account, and for property `popular-products-cd.instawp.xyz`
* Install WooCommerce
* Enable `keyMetrics` feature flag and add `Most popular products by pageviews` key metric widget
* There should be data collected for 3 products added on that demo site

## Changelog entry

* Update the "Most popular products by pageviews" Key Metric tile to utilize a custom dimension in Analytics.

**Comment by tofumatt** [comment]
ACs here are good, moving to IB 👍🏻

**Comment by techanvil** [comment]
Hey @zutigrm, this IB is off to a good start. A few points:

- `googlesitekit_post_type` should be added to `requiredCustomDimensions` in the [definition for `KM_ANALYTICS_POPULAR_PRODUCTS`](https://github.com/google/site-kit-wp/blob/c9bd883549c11655029fae20f2ac5f3b4beb159f/assets/js/components/KeyMetrics/key-metrics-widgets.js#L140).
  - This will result in CD being present in `availableCustomDimensions`, so [this condition](https://github.com/google/site-kit-wp/blob/00aa54ff148d37648ff6c7cf3189a8dbafefbf6f/includes/Modules/Analytics_4.php#L1411) will return true. As a result your specced changes to `get_custom_dimensions_data` won't be needed.
- The `productPostType` field should only be added to the inline base data if the determined post type exists and is public, i.e. [this check]( https://github.com/google/site-kit-wp/blob/c9bd883549c11655029fae20f2ac5f3b4beb159f/includes/Core/Assets/Assets.php#L1088) should be effectively retained.
- The current conditions that check `productBasePaths` should be updated to instead check for the presence of `productPostType` to ensure the widget is not shown unconditionally ([here](https://github.com/google/site-kit-wp/blob/c9bd883549c11655029fae20f2ac5f3b4beb159f/assets/js/components/KeyMetrics/key-metrics-widgets.js#L149), and the usage of [this](https://github.com/google/site-kit-wp/blob/c9bd883549c11655029fae20f2ac5f3b4beb159f/assets/js/modules/analytics-4/components/widgets/PopularProductsWidget.js#L118) variable).
- It's worth mentioning that we're filtering by the `customEvent:googlesitekit_post_type` dimension in the update to `reportOptions`, for the sake of clarity.

**Comment by techanvil** [comment]
Hi @zutigrm, thanks for the update. However, I think you have slightly missed the point I made with regard to `requiredCustomDimensions`, `availableCustomDimensions` and `get_custom_dimensions_data()`.

Adding `googlesitekit_post_type` to `requiredCustomDimensions` for the widget will ensure that once the widget has been selected, the logic for creating/syncing the custom dimensions will run, with the result being that `googlesitekit_post_type` will be added to `availableCustomDimensions` once it's created.

As a result of this, the [existing logic](https://github.com/google/site-kit-wp/blob/69b5e0d6fe7857adb814e92ee5c65bdd20401940/includes/Modules/Analytics_4.php#L1411-L1413) in `get_custom_dimensions_data()` should ensure that `googlesitekit_post_type` is tracked for _all_ post types, including whatever the product post type is, and so no changes to `get_custom_dimensions_data()` should be needed as part of this issue.

Following on from this, the suggested refactor to extract `get_custom_dimensions_data()` to a new class would be pretty out of scope here, and I don't think it's really all that worthwhile to extract `check_for_product_post_type()` (or whatever we call this, see below) to a new class in isolation. Therefore, I'd suggest removing the suggestion for a refactor from this issue. You could always create a separate tech-debt/tidyup issue to propose this refactoring. Although with the Single Analytics Module epic coming up, it might well be worth holding fire to see how the Analytics module looks once that's complete.

Regarding the proposed `hasProductPostType` property seems a bit superfluous, all we really need to do is to conditionally include `productPostType`, and we can simply check if it exists or not in the client-side conditions. Therefore I'd suggest removing `hasProductPostType` from the IB to avoid unnecessary proliferation of inline properties.

One last and very minor point, I'd suggest the slightly terser name `has_product_post_type()` instead of  `check_for_product_post_type()`, as the wording "check for" just feels a touch verbose and possibly redundant, although YMMV :)

**Comment by zutigrm** [comment]
@techanvil Thanks for the feedback.

> Adding googlesitekit_post_type to requiredCustomDimensions for the widget will ensure that once the widget has been selected, the logic for creating/syncing the custom dimensions will run, with the result being that googlesitekit_post_type will be added to availableCustomDimensions once it's created.

> As a result of this, the [existing logic](https://github.com/google/site-kit-wp/blob/69b5e0d6fe7857adb814e92ee5c65bdd20401940/includes/Modules/Analytics_4.php#L1411-L1413) in get_custom_dimensions_data() should ensure that googlesitekit_post_type is tracked for all post types, including whatever the product post type is, and so no changes to get_custom_dimensions_data() should be needed as part of this issue.

I understood that part, but I put the additional condition for is singular and filter with default post type of product, as per AC, we are making this integration specifically for product, and CPT type should be filterable, it should not be applied to all post types. So I proposed additional check to confirm that `product` or other filtered string, exists as post type and we are currently on that CPT.

If no changes are made in `get_custom_dimensions_data`, we do not know which post type is used for product, and users/or plugins cannot filter their product post type name in. And we don't know what string we are filtering in the custom dimensions, beside having hardcoded `product` as value.

> Following on from this, the suggested refactor to extract get_custom_dimensions_data() to a new class would be pretty out of scope here, and I don't think it's really all that worthwhile to extract check_for_product_post_type() (or whatever we call this, see below) to a new class in isolation. Therefore, I'd suggest removing the suggestion for a refactor from this issue.

Makes sense, will remove that part. Thanks.

> Regarding the proposed hasProductPostType property seems a bit superfluous, all we really need to do is to conditionally include productPostType, and we can simply check if it exists or not in the client-side conditions. Therefore I'd suggest removing hasProductPostType from the IB to avoid unnecessary proliferation of inline properties

`hasProductPostType` is needed to determine if filterable "product" post type exists, as `productPostType` is only used to carry the name of product post type - which is "product" by default, and can be filtered by users/plugins. We need both values to determine first, what is the product CPT name, and secondly to know if that CPT is present. We could technically use `productPostType` to carry either CPT name, or false if that CPT is not hooked, but that would beat the goal of AC, as `productPostType` needs to only carry static string which should be filterable, so we need to have both of these variables handling different value types.

**Comment by techanvil** [comment]
Hi @zutigrm, thanks for clarifying your previous update.

> I understood that part, but I put the additional condition for is singular and filter with default post type of product, as per AC, we are making this integration specifically for product, and CPT type should be filterable, it should not be applied to all post types. So I proposed additional check to confirm that `product` or other filtered string, exists as post type and we are currently on that CPT.
> 
> If no changes are made in `get_custom_dimensions_data`, we do not know which post type is used for product, and users/or plugins cannot filter their product post type name in. And we don't know what string we are filtering in the custom dimensions, beside having hardcoded `product` as value.

Please note, at the moment we already do check for `is_singular()` at the [top](https://github.com/google/site-kit-wp/blob/69b5e0d6fe7857adb814e92ee5c65bdd20401940/includes/Modules/Analytics_4.php#L1390-L1392) of `get_custom_dimensions_data()` and bail if it's not true.

Furthermore - we don't need to do any filtering of the post type at this point. We simply want to track what the actual post type is for all singular pages.

The intent of the filter is to allow users to customise what post type is checked for when we query the GA4 API for the report. So, the only point where we need to apply that filter is where we're adding the `productPostType` property to the inline base data.

As a result, my original point still stands - we should not need to make any changes to `get_custom_dimensions_data()` here.

https://github.com/google/site-kit-wp/pull/7661#discussion_r1344614038

> `hasProductPostType` is needed to determine if filterable "product" post type exists, as `productPostType` is only used to carry the name of product post type - which is "product" by default, and can be filtered by users/plugins. We need both values to determine first, what is the product CPT name, and secondly to know if that CPT is present. We could technically use `productPostType` to carry either CPT name, or false if that CPT is not hooked, but that would beat the goal of AC, as `productPostType` needs to only carry static string which should be filterable, so we need to have both of these variables handling different value types.

I still don't see that we need the separate values here. As mentioned above, `productPostType` should contain the _filtered_ value, and it will only be needed on the client-side if the post exists and is public. Therefore we should only include it when that condition is true, and we can avoid the extra property. We could still add a `hasProductPostType()` selector, but this should return derived state with `productPostType` being the source of truth. I'm reminded here of this [previous discussion](https://github.com/google/site-kit-wp/pull/7661#discussion_r1344614038) where some similar ground was covered.

Incidentally we don't tend to use names like `getHasFoo()`, rather just `hasFoo()`.

**Comment by zutigrm** [comment]
@techanvil Thanks, I see what you mean.

> Furthermore - we don't need to do any filtering of the post type at this point. We simply want to track what the actual post type is for all singular pages.

> The intent of the filter is to allow users to customise what post type is checked for when we query the GA4 API for the report. So, the only point where we need to apply that filter is where we're adding the productPostType property to the inline base data.

Got it, you are correct. Indeed `productPostType` should be filterable only in inline data in this scenario. It's clear to me now.

> I still don't see that we need the separate values here. As mentioned above, productPostType should contain the filtered value, and it will only be needed on the client-side if the post exists and is public. Therefore we should only include it when that condition is true, and we can avoid the extra property. 

Yes it makes sense now, thanks.

**Comment by techanvil** [comment]
Thanks @zutigrm! A few additional points:

- We should not add `productPostType` in the `inline_custom_dimensions_data()` method. This property may have utility outside of the Analytics module, and we should add it in the `Assets` class, replacing the `productBasePaths` property. Also, note that adding it in `inline_custom_dimensions_data()` would have the effect of adding the property to `_googlesitekitModulesData`, rather than `_googlesitekitBaseData` which is where we want it.
- As the specced `has_product_post_type()` method actually returns the product type, it should be named `get_product_post_type()`. It should return `null` instead of `false` to avoid returning multiple types.
- The filter name should just be `googlesitekit_product_post_type` as the product post type is standalone concept and not tied to custom dimensions.
- I find the phrasing "Re-use this filter" a bit confusing and would suggest rewording the relevant point along these lines:

> Apply a filter with the hook name `googlesitekit_product_post_type`, with the initial value being the string `product` (see [here](https://github.com/google/site-kit-wp/blob/00aa54ff148d37648ff6c7cf3189a8dbafefbf6f/includes/Modules/Analytics_4.php#L1406)  for an example), and assign the result to a new variable, say `$product_post_type`
- The last subpoint about updating `reportOptions` is also a little confusing and I'd suggest rewriting the overall point along these lines:

> * Update `reportOptions`, 
>    * Add a `dimensionFilter` for the `customEvent:googlesitekit_post_type` dimension
>    * Use `string` for `filterType`
>    * Use `EXACT` for `matchType`
>    * And `productPostType` as a value

- We will also need to add `googlesitekit_post_type` to the list of [valid dimensions in the GA4 `Request` class](https://github.com/google/site-kit-wp/blob/d2c526086329e40e04eb875d1a1663e244e1c23f/includes/Modules/Analytics_4/Report/Request.php#L273-L289).

**Comment by zutigrm** [comment]
Thanks for the feedback @techanvil. I updated the IB

**Comment by techanvil** [comment]
Thanks @zutigrm! This LGTM. IB :white_check_mark:

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Verified that if WooCommerce plugin is enabled and "Most popular products by pageviews" key metric widget is selected then data is showing for the 3 products.
- Verified that if WooCommerce plugin is not enabled then widget showing notice that no products detected on your site.

**Note** : Will check tomorrow to make sure that page views numbers are updating after 24 hours.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/f6fa1991-493b-49c5-8eaf-59c5f3a409d4)

![image](https://github.com/google/site-kit-wp/assets/94359491/6676d6de-3601-4698-b944-e2200428e601)

</details>
